corehq/apps/auditcare/utils/migration.py

```diff
@@ -0,0 +1,94 @@
+from datetime import datetime, timedelta
+
+from django.core.cache import cache
+
+from dimagi.utils.dates import force_to_datetime
+
+from corehq.apps.auditcare.models import AuditcareMigrationMeta
+
+INITIAL_START_DATE = datetime(2013, 1, 1)
+
+
+class AuditCareMigrationUtil():
+
+    def __init__(self):
+        self.start_key = "auditcare_migration_2021_next_batch_time"
+        self.start_lock_key = "auditcare_migration_batch_lock"
+
+    def get_next_batch_start(self):
+        return cache.get(self.start_key)
+
+    def generate_batches(self, worker_count, batch_by):
+        batches = []
+        with cache.lock(self.start_lock_key, timeout=10):
+            start_datetime = self.get_next_batch_start()
+            if not start_datetime:
+                # for the first call
+                start_datetime = INITIAL_START_DATE
+
+            start_time = _get_formatted_start_time(start_datetime, batch_by)
+            end_time = None
+
+            for index in range(worker_count):
+                end_time = _get_end_time(start_time, batch_by)
+                batches.append([start_time, end_time])
+                start_time = end_time
+            self.set_next_batch_start(end_time)
+            if end_time > datetime.now():
+                print("Migration successfully done")
```

Comment:
It looks like this is iterating from the oldest to the newest records. I think it would be better to iterate the other direction since more recent records are likely to be more valuable than older ones. Do you agree?

Reply:
Going either way would have worked for us. There are two reasons why I chose to start from oldest.

- We have very few records in past and they would be real quick to migrate. I was not sure if the data structure/properties have changed. So chances of errors would be high there. Would be easy to full proof the command.
- We already have code in place which divides the query based on time. So even if we are copying the latest records first they would not be that useful for us. And before using them we have to ensure that all the batches have been completed and no batch has errored.

Reply:
> We already have code in place which divides the query based on time.

This is true, but if we store the dividing date in the database we could move/update it as records are migrated. This is assuming the migration will take a long time, which I expect it will in it's current form. I think two SQL queries per record is going to be pretty slow.

I still think it would be best to iterate from newest to oldest records if it's not too hard.